### ISSUE TITLE
Do not spawn more workers than testcases

### ISSUE THREAD

**Issue by azmeuk** [pytest-dev/pytest-xdist#272]
I switched from a Core i5 (4 cores) to a Ryzen 7 (16 cores). The time pytest-xdist takes to spawn the workers is consequently greater, but it is expected.

Sometimes I run pytest for just several tests: sometimes 1 test, sometimes 1 files with a few tests. I thought it would save some time if pytest-xdist would spawn only the workers needed.

What do you think?

**Comment by RonnyPfannschmidt** [comment]
when xdist starts the workers, it has zero idea how many tests are there, so currently its not implementable

**Comment by Krilivye** [comment]
Should the collect phase start before launching the workers?

**Comment by nicoddemus** [comment]
I guess it would be possible to launch a single worker, collect the tests, and then based on the number of tests spawn the rest of the workers as needed, but this would probably require quite a bit of refactoring to accomplish.

**Comment by RonnyPfannschmidt** [comment]
@Krilivye the master does not run any collection, all collection is on the workers

**Comment by azmeuk** [comment]
Has it been designed willingly like this for technical reasons?

**Comment by nicoddemus** [comment]
Yes, because transferring tests and fixtures over the wire (between workers and the master node) is quite difficult/impossible, so every worker collects all tests and the master just schedules which test to run based on their node id (take a look at [OVERVIEW](https://github.com/pytest-dev/pytest-xdist/blob/master/OVERVIEW.md) for more details).

Some related issues: #270, #271

**Comment by treyturner-nirvana** [comment]
This (somewhat fragile) one-liner can externalize the 'collect only first' approach. Replace `echo 1` with the desired fallback value.

```
NUM_TESTS=$( (pytest --dist no --lf --collect-only \
    | grep -oE 'previous [0-9]+ failures' \
    | grep -oE '[0-9]+') \
    || echo 1 ); \
pytest -n logical --lf --maxprocesses=$NUM_TESTS
```
